hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/snapshot/TestSnapshotDiffManager.java

```diff
@@ -0,0 +1,333 @@
+import org.apache.hadoop.ozone.om.helpers.OmKeyInfo;
+import org.apache.hadoop.ozone.om.helpers.OmVolumeArgs;
+import org.apache.hadoop.ozone.om.helpers.SnapshotDiffJob;
+import org.apache.hadoop.ozone.om.helpers.SnapshotInfo;
+import org.apache.hadoop.ozone.snapshot.SnapshotDiffResponse;
+import org.apache.hadoop.ozone.snapshot.SnapshotDiffResponse.JobStatus;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.hadoop.security.authentication.client.AuthenticationException;
+import org.apache.hadoop.util.ClosableIterator;
+import org.apache.ozone.test.GenericTestUtils;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+import org.rocksdb.RocksDBException;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.UUID;
+import java.util.concurrent.ThreadLocalRandom;
+
+import static org.apache.hadoop.hdds.protocol.proto.HddsProtos.ReplicationFactor.ONE;
+import static org.apache.hadoop.ozone.om.OmSnapshotManager.DELIMITER;
+
+/**
+ * Tests for {@link SnapshotDiffManager}.
+ */
+public class TestSnapshotDiffManager {
+
+  private static final String VOLUME = "vol";
+  private static final String BUCKET = "bucket";
+
+  private static File metaDir;
+  private static OzoneManager ozoneManager;
+  private static OMMetadataManager omMetadataManager;
+  private static SnapshotDiffManager snapshotDiffManager;
+  private static PersistentMap<String, SnapshotDiffJob> snapDiffJobTable;
+
+  @BeforeAll
+  public static void init() throws AuthenticationException,
+      IOException, RocksDBException {
+    metaDir = GenericTestUtils.getRandomizedTestDir();
+    if (!metaDir.exists()) {
+      Assertions.assertTrue(metaDir.mkdirs());
+    }
+    OzoneConfiguration conf = new OzoneConfiguration();
+    conf.set(OzoneConfigKeys.OZONE_METADATA_DIRS,
+        metaDir.getAbsolutePath());
+
+    OmTestManagers omTestManagers = new OmTestManagers(conf);
+    ozoneManager = omTestManagers.getOzoneManager();
+    omMetadataManager = omTestManagers.getMetadataManager();
+
+    snapshotDiffManager = ozoneManager
+        .getOmSnapshotManager().getSnapshotDiffManager();
+    snapDiffJobTable = snapshotDiffManager.getSnapDiffJobTable();
+
+    createVolumeAndBucket();
+  }
+
+  @AfterAll
+  public static void cleanUp() {
+    FileUtil.fullyDelete(metaDir);
+  }
+
+  /**
+   * Clear the SnapshotDiffTable before every test run.
+   */
+  @BeforeEach
+  public void setUp() {
+    ClosableIterator<Map.Entry<String, SnapshotDiffJob>>
+        iterator = snapDiffJobTable.iterator();
+    while (iterator.hasNext()) {
+      String key = iterator.next().getKey();
+      snapDiffJobTable.remove(key);
+    }
+  }
+
+  @ParameterizedTest
```

Comment:
nit:
```suggestion
  private static Stream<Arguments> listSnapshotDiffJobsScenarios() {
    return Stream.of(
        Arguments.of("queued", false, false),
        Arguments.of("done", false, false),
        Arguments.of("in_progress", false, true),
        Arguments.of("invalid", false, false),
        Arguments.of("", false, false),
        Arguments.of("queued", true, true),
        Arguments.of("done", true, true),
        Arguments.of("in_progress", true, true),
        Arguments.of("invalid", true, true),
        Arguments.of("", true, true)
    );
  }

  @ParameterizedTest
  @MethodSource("listSnapshotDiffJobsScenarios")
  public void testListSnapshotDiffJobs(String jobStatus,
                                       boolean listAll,
                                       boolean containsJob) throws IOException {

    String fromSnapshotName = "snap-" + RandomStringUtils.randomNumeric(5);
    String toSnapshotName = "snap-" + RandomStringUtils.randomNumeric(5);
    String fromSnapshotId = UUID.randomUUID().toString();
    String toSnapshotId = UUID.randomUUID().toString();
    String diffJobKey = fromSnapshotId + DELIMITER + toSnapshotId;

    setUpKeysAndSnapshots(fromSnapshotName, toSnapshotName,
        fromSnapshotId, toSnapshotId);

    SnapshotDiffJob diffJob = snapDiffJobTable.get(diffJobKey);
    Assertions.assertNull(diffJob);

    // There are no jobs in the table, therefore
    // the response list should be empty.
    List<SnapshotDiffJob> jobList = snapshotDiffManager
        .getSnapshotDiffJobList(VOLUME, BUCKET, jobStatus, listAll);
    Assertions.assertTrue(jobList.isEmpty());

    
    SnapshotDiffResponse snapshotDiffResponse = snapshotDiffManager
        .getSnapshotDiffReport(VOLUME, BUCKET,
            fromSnapshotName, toSnapshotName,
            0, 0, false);
    
    Assertions.assertEquals(JobStatus.IN_PROGRESS,
        snapshotDiffResponse.getJobStatus());

    diffJob = snapDiffJobTable.get(diffJobKey);
    Assertions.assertNotNull(diffJob);
    Assertions.assertEquals(JobStatus.IN_PROGRESS,
        diffJob.getStatus());

    jobList = snapshotDiffManager
        .getSnapshotDiffJobList(VOLUME, BUCKET, jobStatus, listAll);

    if (containsJob) {
      Assertions.assertTrue(jobList.contains(diffJob));
    } else {
      Assertions.assertTrue(jobList.isEmpty());
    }
  }
```

On a side note, there is no need to add straight forward comments like  `// Submit a job.` or `// Response should be IN_PROGRESS.`